assignment/doxygen/man/man3/IOManager.3
.TH "IOManager" 3 "Sun Dec 10 2017" "Heat conduction equation" \" -*- nroff -*-
.ad l
.nh
.SH NAME
IOManager \- An input/output manager class to handle plot exportations and future implementations of input handling\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <iomanager\&.h>\fP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBIOManager\fP ()"
.br
.RI "\fIDefault constructor\&. \fP"
.ti -1c
.RI "void \fBexport_outputs\fP (\fBMethod\fP *analytical, std::vector< \fBMethod\fP * > methods)"
.br
.RI "\fIExports outputs regarding plots images and error tables for each computed solution, comparing them to the analytical solution\&. \fP"
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "bool \fBcreate_output_dir\fP ()"
.br
.RI "\fI\fBMethod\fP to create ouput folder if the folder does not exist\&. \fP"
.ti -1c
.RI "void \fBplot_solutions\fP (std::string output_name, \fBMethod\fP *analytical, \fBMethod\fP *method)"
.br
.RI "\fIExports a plot chart that compares the analytical solution to any other solution using gnuplot\&. \fP"
.ti -1c
.RI "void \fBplot_laasonen_times\fP ()"
.br
.RI "\fIExports a plot with \fBLaasonen\fP delta t variation computational times\&. \fP"
.ti -1c
.RI "void \fBplot_default_deltat_times\fP ()"
.br
.RI "\fIExports a plot with four methods computational times\&. \fP"
.ti -1c
.RI "void \fBerror_tables\fP (std::string output_name, std::vector< \fBMethod\fP * > method)"
.br
.RI "\fIExports a plot that compares the norms of each solution\&. \fP"
.ti -1c
.RI "std::string \fBdouble_to_string\fP (int precision, double value)"
.br
.RI "\fIConverts a double to a string with a precison of 2 decimal places\&. \fP"
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "std::string \fBoutput_path\fP"
.br
.RI "\fIPrivate string output_path\&. \fP"
.ti -1c
.RI "std::vector< double > \fBlaasonen_times\fP"
.br
.RI "\fIPrivate \fBVector\fP laasonen_times\&. \fP"
.ti -1c
.RI "std::vector< double > \fBdefault_deltat_times\fP"
.br
.RI "\fIPrivate \fBVector\fP default_deltat_times\&. \fP"
.in -1c
.SH "Detailed Description"
.PP 
An input/output manager class to handle plot exportations and future implementations of input handling\&. 

The \fBIOManager\fP class provides: 
.br
-plot method which compares the analytical solution with a set of given methods, ploting them with a custom configuration using gnuplot 
.SH "Constructor & Destructor Documentation"
.PP 
.SS "IOManager::IOManager ()"

.PP
Default constructor\&. Initialize an \fBIOManager\fP object\&. 
.SH "Member Function Documentation"
.PP 
.SS "bool IOManager::create_output_dir ()\fC [private]\fP"

.PP
\fBMethod\fP to create ouput folder if the folder does not exist\&. 
.PP
\fBReturns:\fP
.RS 4
bool\&. true if successfull, false if not 
.RE
.PP

.SS "std::string IOManager::double_to_string (int precision, double value)\fC [private]\fP"

.PP
Converts a double to a string with a precison of 2 decimal places\&. 
.PP
\fBParameters:\fP
.RS 4
\fIdouble\fP value Number to be converted 
.br
\fIint\fP precision Precision to have 
.RE
.PP
\fBReturns:\fP
.RS 4
string\&. String containing the converted number 
.RE
.PP

.SS "void IOManager::error_tables (std::string output_name, std::vector< \fBMethod\fP * > method)\fC [private]\fP"

.PP
Exports a plot that compares the norms of each solution\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP output_name File name to be exported 
.br
\fIvector<Method*>\fP vector of methods to plot the second norm 
.RE
.PP

.SS "void IOManager::export_outputs (\fBMethod\fP * analytical, std::vector< \fBMethod\fP * > methods)"

.PP
Exports outputs regarding plots images and error tables for each computed solution, comparing them to the analytical solution\&. 
.PP
\fBParameters:\fP
.RS 4
\fIMethod*\fP analytical The analytical solution 
.br
\fIvector<Method*>\fP methods \fBVector\fP containing the solutions 
.RE
.PP

.SS "void IOManager::plot_default_deltat_times ()\fC [private]\fP"

.PP
Exports a plot with four methods computational times\&. 
.SS "void IOManager::plot_laasonen_times ()\fC [private]\fP"

.PP
Exports a plot with \fBLaasonen\fP delta t variation computational times\&. 
.SS "void IOManager::plot_solutions (std::string output_name, \fBMethod\fP * analytical, \fBMethod\fP * method)\fC [private]\fP"

.PP
Exports a plot chart that compares the analytical solution to any other solution using gnuplot\&. 
.PP
\fBParameters:\fP
.RS 4
\fIstring\fP output_name File name to be exported 
.br
\fIMethod*\fP analytical The analytical solution 
.br
\fIMethod*\fP method Any method solution 
.RE
.PP

.SH "Member Data Documentation"
.PP 
.SS "std::vector<double> IOManager::default_deltat_times\fC [private]\fP"

.PP
Private \fBVector\fP default_deltat_times\&. Contains the computation time of each method solution, with a time step of 0\&.01\&. 
.SS "std::vector<double> IOManager::laasonen_times\fC [private]\fP"

.PP
Private \fBVector\fP laasonen_times\&. Contains the computation time of each laasonen solution, with a different time step\&. 
.SS "std::string IOManager::output_path\fC [private]\fP"

.PP
Private string output_path\&. Contains the ouput directory path name\&. 

.SH "Author"
.PP 
Generated automatically by Doxygen for Heat conduction equation from the source code\&.


reports/src/io/iomanager.cpp
#include "iomanager.h"

// CONSTRUCTORS
/*=
 * Default constructor
 */

IOManager::IOManager() {
	output_path = OUTPUT_PATH;
}

// PLOT METHODS

/*
* private output method - creates output folder 
*/

bool IOManager::create_output_dir() {
	char answer = '.';

	// if plot folder exists ask if the user wants to replace the previous results
	if(boost::filesystem::exists(output_path)) {
		while (answer != 'y' && answer != 'Y' && answer != 'n' && answer != 'N') {
			std::cout << "Output folder already exists, this program might overwrite some files, procceed? [Y:N] - ";
			answer = std::getchar();
			std::cout << std::endl;
		}
		return answer == 'y' || answer == 'Y' ? true : false;
	} else {
		if(boost::filesystem::create_directory(output_path)) {
			std::cout << "Output folder was created!" << std::endl;
			return true;
		} else {
			std::cout << "Output folder couldn't be created! Not exporting outputs." << std::endl;
			return false;
		}
	}
}

/*
* public output method - iterates through all the solutions in order to export them in varied formats
*/

void IOManager::export_outputs(Method * analytical, std::vector<Method*> methods) {
	if (!create_output_dir()) return;
	std::cout << "Exporting outputs to " << boost::filesystem::canonical(output_path, ".") << std::endl;
	std::string name, output_name, deltat_string;
	for (unsigned int index = 0; index < methods.size(); index++) {
		name = (*methods[index]).get_name();
		deltat_string = name == LAASONEN ? double_to_string(3, methods[index]->get_deltat()) : double_to_string(2, methods[index]->get_deltat());
		output_name = output_path + '/' + name;
		if (name == LAASONEN) {
			output_name += "dt=" + deltat_string;
			laasonen_times.push_back(methods[index]->get_computational_time());
		} 
		if (! (name == LAASONEN && methods[index]->get_deltat() != 0.01) && name != RICHARDSON) {
			default_deltat_times.push_back(methods[index]->get_computational_time());
		}
		std::cout << "Exporting " << name << " method outputs... ";
		plot_solutions(output_name, analytical, methods[index]);
		std::cout << "Finished!" << std::endl;
	}
	std::vector<Method*> error_vector(methods.begin() + 1, methods.begin() + 4);
	error_tables(output_name, error_vector);
	plot_default_deltat_times();
	plot_laasonen_times();
}

/*
* private plot method - Exports a plot chart which compares the analytical solution with a given solution
*/

void IOManager::plot_solutions(std::string output_name, Method * analytical, Method * method) {
	// Object to export plots
	Gnuplot gp;

	// methods solutions
	Matrix analytical_matrix = analytical->get_solution(), method_matrix = method->get_solution();
	unsigned int rows = method_matrix.getNrows();
	unsigned int cols = method_matrix.getNcols();
	double time;
	std::string time_str, name = method->get_name();

	// defining gnuplot configuration with the correct syntax
	gp << "set key on box; set tics scale 0; set border 3; set ylabel \"Temperature [F]\";set xlabel \"x [ft]\"; set yrange [90:310]; set term png; set xtics (\"0\" 0, \"0.5\"" << cols / 2 << ", \"1\"" << cols - 1 << ")\n";
	for (unsigned int index = 1; index < rows; index++) {
		time = (double)index / 10.0;
		time_str = double_to_string(1, time);

		gp << "set output \"" << output_name << "t="  << time_str;
		gp << ".png\";\n";
		gp << "plot" << gp.file1d(analytical_matrix[index]) << "with lines title \"Analytical\" lw 2 lt rgb \"red\","
			<< gp.file1d(method_matrix[index]) << "with points title \"" << name << "\" pt 17 ps 1 lw 1" << std::endl;
	}
}

/*
* private plot method - Exports a plot chart with laasonen solutions computational times
*/

void IOManager::plot_laasonen_times() {
	// Object to export plots
	Gnuplot gp;

	gp << "set tics scale 0; set border 3; set style line 1 lc rgb '#0060ad' lt 1 lw 2 pt 7 pi -1 ps 1.5; set clip two; set ylabel \"Computational Time [ms]\";set xlabel \"Delta t [h]\"; set term png; set xtics (\"0.01\" 0, \"0.025\" 1, \"0.05\" 2, \"0.1\" 3)\n";
	gp << "set output \"" << output_path << "/laasonen_times.png\";\n";
	gp << "plot" << gp.file1d(laasonen_times) << " notitle with linespoint ls 1" << std::endl;
}

/*
* private plot method - Exports a plot chart with solutions computational times
*/

void IOManager::plot_default_deltat_times() {
	// Object to export plots
	Gnuplot gp;

	gp << "set tics scale 0; set border 3; set style line 1 lc rgb '#0060ad' lt 1 lw 2 pt 7 pi -1 ps 1.5; set clip two; set ylabel \"Computational Time [ms]\";set xlabel \"Delta t [h]\"; set term png; set xtics (\"Dufort-Frankel\" 0, \"Crank-Nicholson\" 1, \"Laasonen\" 2)\n";
	gp << "set output \"" << output_path << "/default_deltat_times.png\";\n";
	gp << "plot" << gp.file1d(default_deltat_times) << " notitle with linespoint ls 1" << std::endl;
}

/*
* private table method - Exports an error table to a .lsx file which compares the analytical solution with a given solution
*/
void IOManager::error_tables(std::string output_name, std::vector<Method*> methods) {

	// Object to export plots
	Gnuplot gp;
	std::vector<double> norms;

	for (int i = 0; i < methods.size(); i++) {
		norms.push_back(methods[i]->get_two_norm());
	}
	std::swap(norms[1],norms[2]);


	gp << "set tics scale 0; set border 3; set style line 1 lc rgb '#FFA500' lt 1 lw 2 pt 7 pi -1 ps 1.5; set clip two; set ylabel \"2nd Norm\";set xlabel \"\"; set term png; set xtics (\"Dufort-Frankel\" 0, \"Laasonen\" 1, \"Crank Nicholson\" 2)\n";
	gp << "set output \"" << output_path << "/norms.png\";\n";
	gp << "plot" << gp.file1d(norms) << " notitle with linespoint ls 1" << std::endl;
}

// AUX METHODS

/*
* aux method - Converts a double into a string
*/

std::string IOManager::double_to_string(int precision, double value) {
	std::stringstream stream;
	stream << std::fixed << std::setprecision(precision) << value;
	return stream.str();
}